Modify Rule S2260: Add Dart language (#4158)
SonarSource/rspec: rules/S2260/dart/rule.adoc

== Why is this an issue?

> When the analysis succeeds, it doesn't mean that the analyzer was able to understand all the analyzed code. If the analyzer fails to parse some parts of your code, it will ignore them during the analysis. This rule will help you track these parsing failures.

Reviewer: ```suggestion
When the analysis succeeds, it doesn't mean that the analyzer was able to understand all the analyzed code. If the analyzer fails on some parts of your code, it will ignore them during the analysis. This rule will help you track these analysis failures.
```

> There are many reasons why parsing failures can happen, here are the common ones:

Reviewer: ```suggestion
There are many reasons why analysis failures can happen, here are the common ones:
```

> * The code contains compile-time error(s)

Reviewer: ```suggestion
* The code contains compile-time error(s).
```

> * The types weren't resolved correctly (some dependencies are missing or files weren't generated)

Reviewer: ```suggestion
* The types weren't resolved correctly (some dependencies are missing or files weren't generated).
```

* Use of new language features that are not yet supported by our analyzer.

> 

Reviewer: ```suggestion
* Limitation in our analyzer. We are always working on improving this.

```

> How do they impact analysis? We cannot judge without looking at specific examples, as they contain a broad range of types of errors. On our side, we will make sure that you get notified through the analysis logs when they have an impact on the quality of the analysis.

Reviewer: ```suggestion
How do they impact analysis? We cannot judge without looking at specific examples, as they contain a broad range of types of errors.
```

> There are three recommended ways to deal with parsing failures:

Reviewer: ```suggestion
There are three recommended ways to deal with analysis failures:
```

> * Fix compiler errors

Reviewer: ```suggestion
* Fix compiler errors.
```

> * Make sure all referenced generated files were generated before the analysis

Reviewer: ```suggestion
* Make sure all referenced generated files were generated before the analysis.
```

> * If you cannot fix them and the analysis logs don’t state anything explicit about their impact, ignore them by resolving them as "won't fix".

Reviewer: ```suggestion
* If you cannot fix them, let us know through the https://community.sonarsource.com/[Sonar Community forum].
```

=== Noncompliant code example

[source,dart,diff-id=1,diff-type=noncompliant]
----
void fun(int x) {
  print(x) // Noncompliant, missing ';'
}

----

=== Compliant solution

[source,dart,diff-id=1,diff-type=compliant]
----
void fun(int x) {
  print(x);
}

----

=== Noncompliant code example

Another example could be with missing generated classes

[source,dart]
----

> import 'package:generated/my_classes.dart'; // Noncompliant, Target of URI doesn't exist: 'package:generated/my_classes.dart'.

Reviewer: ```suggestion
import 'package:generated/my_classes.dart'; // Noncompliant, Target of URI doesn't exist: 'package:generated/my_classes.dart'.

```
